### PR TITLE
Add doc discussing enumerations versus identifiers

### LINKED ISSUE #596: Add design doc discussing enumerations versus identifiers
There are many areas around ICU4X where we have a set of items that we need to identify.  There are two main strategies that can be employed: enumerations and identifier strings.

Sometimes, it is unclear whether a situation warrants an enumeration or an identifier string.  One such situation came up in the timezone provider, where we were deciding whether to represent time zone variants as an enum or an identifier (see https://github.com/unicode-org/icu4x/pull/512#discussion_r603629750).

Here are the principles I would like to apply across ICU4X:

1. If the set of items is bounded and universal, use an enumeration.
2. Otherwise, use an identifier.
3. If in doubt, use an identifier.

Examples where the choice is clear:

- `Signum` is an enumeration, because there are 4 well-defined items that are universal.
- `Language` is an identifier, because the set of possible languages is subject to change over time as new languages are discovered and invented.
- `MetaZone` and `TimeZone` are identifiers, because time zones are subject to change.

Examples where the choice is a bit less clear:

- `PluralForm` has 6 known variants (zero, one, two, few, many, other), and this list has not grown recently, even as more languages are discovered.  However, the list is not necessarily universal, and it is in theory subject to change.
- `TimeZoneVariant`, mentioned above, has 3 commonly used variants (generic, standard, daylight), but this list is *not* universal, since many countries do not have daylight savings time, and some introduce locale-specific variants such as Ramadan time or market time.

### PR DESCRIPTION
**Pull request by sffc**
Fixes #596

This is relevant to our Unicode properties discussion.  I advocated to represent them with numerical IDs, on the basis that the set of IDs grows over time.  That type of decision-making is represented in this design principles doc, which is long overdue.

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]
This looks great!

**Comment by zbraniecki** [comment]
I feel this document misrepresents the two main value propositions of enumerations over identifiers.

1) Enumerations are incredibly useful for large codebase maintenance. In the exact opposite way from identifiers, enumerations do amazing job at helping engineers update all callsites when the set changes.
2) Enumerations guarantee validity. a String or TinyStr *may* be a valid entity, but it has to be parsed and validated at a cost and attention has to be continuously paid to the API of the new type to ensure that as code evolves no pathway gets creates to produce an invalid identifier inside the new type.

The fact that there is no mention of either of those points makes me question whether I am misguided to see this value.

Can you share your thoughts Shane, Manish?

**Comment by sffc** [comment]
> I feel this document misrepresents the two main value propositions of enumerations over identifiers.

Thanks for sharing more value propositions of enumerations!  This is what I was hoping to get out of this review.

> 1. Enumerations are incredibly useful for large codebase maintenance. In the exact opposite way from identifiers, enumerations do amazing job at helping engineers update all callsites when the set changes.

I agree that this is a nice benefit, and I will update my doc to emphasize it as a reason to use enums in situations where they are otherwise appropriate.

Not that enums have refactoring benefits only if they are not `#[non_exhaustive]`.  I argue in my doc that you should generally use an identifier over a non-exhaustive enum.

> 2. Enumerations guarantee validity. a String or TinyStr _may_ be a valid entity, but it has to be parsed and validated at a cost and attention has to be continuously paid to the API of the new type to ensure that as code evolves no pathway gets creates to produce an invalid identifier inside the new type.

It's true that you need to perform validation in more places with identifiers, but I would counter-argue that you also need to validate enums that are created from sources like data files.

I don't find the "attention has to be continuously paid" argument to be compelling.  The backing data store of an identifier is private, and all constructors on the newtype should perform validation or be named `unchecked`.  I don't see that as being a challenge.

I will add these points to the doc.

File: docs/design/enums_or_ids.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+Enums or Identifiers
```
Comment (echeran) [inline docs/design/enums_or_ids.md:1]: I understand the topic that this is addressing, but the naming and definitions could help.

I think the topic is about representing values -- specifically, when the range of values is finite, and the values form a discrete set.  So, a finite set of discrete values.

I'd say that an entity is a standalone thing. Instead of the word "entity", I would use the word "element" instead to represent an item in a set -- both for the CS and the math definitions of "set".

I agree on the definition of identifier as a space of strings consisting of numbers, letters, etc. that stand in for something. (Some languages define "keyword" as a language-reserved type of identifier with specific semantics, and leave "identifier" for user-defined constructs. Other languages use "symbol" to represent the entire space of such syntactic token strings of keywords+identifiers.)  

But long story short, I don't think "identifier" is the right word to use (as a subcategory of this larger topic, for which enum is one alternative).

Integers and strings are used as examples for what is called "identifier". But I don't think we would use a floating point or a fixed decimal in such examples. So I think the key characteristic is that integers and strings are _unbounded_ discrete values. 

The unbounded nature of integers and strings gets to the heart of the problem that the decision tree questions address -- (are we dealing with discrete values? is the range a finite set?)  what is the size (cardinality) of the set of values? can the set grow? 

"Unbounded discrete value type" isn't concise like "identifier", but they do feel very distinct to me. You could also just say "ints/strings", that sounds fine to me.  The newtype idiom seems like a wrapper around some primitive type, from what I've seen of it so far.

Comment (sffc) [reply docs/design/enums_or_ids.md:1]: Thanks for the feedback.  Responses:

> I understand the topic that this is addressing, but the naming and definitions could help.
> 
> I think the topic is about representing values -- specifically, when the range of values is finite, and the values form a discrete set. So, a finite set of discrete values.

The discrete set is not necessarily finite. In many cases, it may be *de facto* finite but theoretically infinite.

> I'd say that an entity is a standalone thing. Instead of the word "entity", I would use the word "element" instead to represent an item in a set -- both for the CS and the math definitions of "set".

Part of my reasoning in choosing the word "entity" was because I wanted to emphasize that the thing in question has standalone value.  For example, it is meaningful to talk about a certain language or a certain time zone outside the context of talking about the other elements of the set.  I will clarify that in the doc.

> I agree on the definition of identifier as a space of strings consisting of numbers, letters, etc. that stand in for something. (Some languages define "keyword" as a language-reserved type of identifier with specific semantics, and leave "identifier" for user-defined constructs. Other languages use "symbol" to represent the entire space of such syntactic token strings of keywords+identifiers.)
> 
> But long story short, I don't think "identifier" is the right word to use (as a subcategory of this larger topic, for which enum is one alternative).

Part of my inspiration for choosing the terminology I chose is UTS 39:

http://unicode.org/reports/tr39/#Identifier_Characters

From that spec: "Identifiers ("IDs") are strings used in application contexts to refer to specific entities of certain significance in the given application. In a given application, an identifier will map to at most one specific entity."

I will add a reference to the spec in the doc.

> Integers and strings are used as examples for what is called "identifier". But I don't think we would use a floating point or a fixed decimal in such examples. So I think the key characteristic is that integers and strings are _unbounded_ discrete values.
>
> The unbounded nature of integers and strings gets to the heart of the problem that the decision tree questions address -- (are we dealing with discrete values? is the range a finite set?) what is the size (cardinality) of the set of values? can the set grow?
> 
> "Unbounded discrete value type" isn't concise like "identifier", but they do feel very distinct to me. You could also just say "ints/strings", that sounds fine to me. The newtype idiom seems like a wrapper around some primitive type, from what I've seen of it so far.

An identifier is not necessarily unbounded.  There are only 256 possible entries if a u8 is used, for example.  The identifier can also specify its own syntax of valid values.  For example, there are only 26^2 possible ISO 639-1 language subtags.

The reason we would most likely not use a floating point number in this context is because floating point numbers lack precision, and there may be multiple bit patterns for the same value (e.g., NaN has many bit patterns).  I imagine that a fixed decimal could be used as an identifier, but more likely you would use a string containing a fixed decimal, since strings are more widely accepted as a medium of communication between programs.

**Review by Manishearth — DISMISSED** [review]

File: docs/design/enums_or_ids.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+====================
+
+An *entity* is a discrete item in a set. For example, English is an entity: it is a specific instance of a language.
+
+There are two general ways to represent entities: *enumerations* and *identifiers*. An *enumeration* is a fixed, finite list of entities. An *identifier* is a universally understood space of numbers or strings, which may or may not be finite. For example, the identifier `"en"` is a language code according to a standard published by ISO.
+
+This design doc helps explain how we think about entities in ICU4X and when to use enums versus identifiers to represent them.
+
+## Decision Tree
```
Comment (iainireland) [inline docs/design/enums_or_ids.md:10]: To my mind, the key difference between enumerations and the new-type pattern is that Rust enums come with a guarantee that the value you're looking at comes from a known universe of values. If you have an enum-typed value, then Rust guarantees that it is valid and the rest of the codebase understands how to deal with it. The newtype case does not have the same guarantee.

This is both good and bad. The good part, as we discussed in the meeting, is that it lets older code work with newer data. The bad part is that it makes it harder for the compiler to help you. For example, if I accidentally write `Script::Laitn`, it's a compiler error; if I accidentally write `Script("Laitn")`, it's a bold new development in linguistics. Similarly, if I have a `#[non_exhaustive]` enum, I can still depend on the compiler to verify that matches _inside_ the crate are exhaustive, whereas the newtype case requires a wildcard pattern even in our own code. 

(From Shane's response to Zibi:
> I don't find the "attention has to be continuously paid" argument to be compelling. The backing data store of an identifier is private, and all constructors on the newtype should perform validation or be named unchecked. I don't see that as being a challenge.

The difference between enums and newtypes here is that Rust won't let you construct an enum without validation (or `unsafe`), whereas rules about newtype constructors can only be enforced by human vigilance.)

On this basis, I think that enums are more ergonomic / less error-prone. That's not to say that I think we should use enums everywhere; as I mentioned in the meeting last week, I find the old-code/new-data argument compelling. But I think we should make sure we are getting something by giving up ergonomics.

All of that is to say: I think this decision tree leans a little too heavily towards identifiers. For example, `GeneralCategory` has more than 20 possible values, but no new values will ever be added, so there's no reason not to make it an enum. I would flip the questions around:

1. How often are new entities added to the set?
   - Set is fixed or grows rarely: Use an enum.
   - Spontaneously: Use a newtype.
   - By an authority, at least once every few years: Continue to the next question.

2. How consistent is the handling of each entity? How hard is it to write code to handle unknown entities? Do we need to know anything about the entity in our own code, or can we pass it straight through to the data provider?
   - Very consistent: Use a newtype.
   - Inconsistent: Use an enum. 

Note that a Script newtype doesn't do anything with our current implementation of enumerated properties, because script data is accessed via ResourceKey (eg "sc=Latn"), and there's no ResourceKey defined for scripts we don't know about. If we want to be able to handle new Script data with old code, then we need to overhaul the data provider.

File: docs/design/enums_or_ids.md (hunk 1 above)
Comment (echeran) [reply docs/design/enums_or_ids.md:1]: Ah, okay, I think I understand better now, the space of string values that you are considering are those that would also qualify as UTS 39 identifiers. And you are using the word "identifier" in that sense. That seems reasonable to me.

And I see the point about how the space of identifiers are not necessarily infinite, but for our purposes (memory/data size considerations), often times it might be.

The main point I was making about finite/bounded vs. unbounded is that enums are finite and discrete, so to have the identifiers be an alternative for situations that the decision tree describes, then for whatever identity / value under consideration, the set of valid values is always something that is finite and discrete. That set of values may expand, and that's the point of the doc.

But I think my previous comment went off on a tangent b/c I didn't understand what "identifier" was representing.

File: docs/design/enums_or_ids.md (hunk 2 above)
Comment (echeran) [reply docs/design/enums_or_ids.md:10]: I agree with this. I would add on to the note at the end that planning to support the representation of new unknown values doesn't necessarily confer any new abilities (ex: as you said, can't load UnicodeSet data for new property values we haven't seen). But we can at least prevent blowing up in the use case of returning property values where, for example, the value we deserialize from the CodePointTrie of newer data is outside the defined set of variants in our older enum.

Maybe what that means is that avoiding that ungraceful error state is the limit of what we can guarantee as the benefit here for identifiers over enums?

**Review by echeran — DISMISSED** [review]

File: docs/design/enums_or_ids.md (hunk 2 above)
Comment (iainireland) [reply docs/design/enums_or_ids.md:10]: One more data point: I'm currently working on converting the existing Script enum into an identifier and it's ... kind of unpleasant?

In particular, we run into this problem I mentioned above:
> Similarly, if I have a #[non_exhaustive] enum, I can still depend on the compiler to verify that matches inside the crate are exhaustive, whereas the newtype case requires a wildcard pattern even in our own code.

If `Script` is just `pub struct Script(pub(crate) u16)` with a bunch of associated constants, then the match in [`get_script_val_set`](https://github.com/unicode-org/icu4x/blob/adf02273511847c7b3422a0916e05d8ac7c236ec/components/uniset/src/props.rs#L550-L556) is no longer exhaustive, so we have to figure out what to do if we're given a `Script` wrapping an identifier we don't know about. Meanwhile, to create the associated constants, I've ended up writing this, which feels like I'm holding the hammer backwards:
```

Comment (sffc) [reply docs/design/enums_or_ids.md:10]: For example, we don't have constants for all the language codes we support. You create then via their language code as a string.

Why should scripts be different than language codes in this regard?

Comment (sffc) [reply docs/design/enums_or_ids.md:10]: In other words, in cases where it makes sense to use identifiers, there shouldn't necessarily be a comprehensive list of all of the identifiers in code.  All that we need is a data-driven map from the 4-letter codes to the numbers, which is backed by the data provider.  ICU4C updates its header files to add in the new entries with each release; I would like if we didn't have to do that and could instead take all mutable data and stick it in the data provider.

Comment (iainireland) [reply docs/design/enums_or_ids.md:10]: I think two-letter language codes have more developer mindshare as a canonical identifier than 4-letter script codes. For example, the ECMA spec makes both the long and short names available for scripts, but defines the _long_ name as the "canonical property value". People are less likely to have the 4-letter code memorized, so it's nice if there's a convenient list of human-readable values (which can also be used in eg IDE autocompletion).

Also, the implicit values (Common, Inherited, Unknown) that developers are most likely to need to hardcode have particularly awful names (Zyyy, Zinh/Qaai, Zzzz). As a developer, I would much prefer to write (or read) `Script::Common` rather than `Script::from_code("Zyyy").unwrap()`.

Basically, I think we provide a better developer experience if there's a list of identifiers available, and I'm not convinced that having to recompile for a new release is a sufficient problem. Given that ICU4C currently doesn't support swapping out data without recompiling, nobody's existing workflow will be blocked if ICU4X doesn't support that use case.

Comment (iainireland) [reply docs/design/enums_or_ids.md:10]: > Also, the implicit values (Common, Inherited, Unknown) that developers are most likely to need to hardcode have particularly awful names (Zyyy, Zinh/Qaai, Zzzz). As a developer, I would much prefer to write (or read) Script::Common rather than Script::from_code("Zyyy").unwrap().

Actually, this might be a good question for the checklist: "Do some values have a different meaning / require special handling? If yes, lean towards enum." Country codes are a flat list, with no special cases; scripts have special cases, which makes it more likely that people will need to refer to those special cases directly, which makes the ergonomic issues more salient.

Comment (sffc) [reply docs/design/enums_or_ids.md:10]: > I think two-letter language codes have more developer mindshare as a canonical identifier than 4-letter script codes.

Not sure if I agree with that.  Anyone familiar with BCP-47 already knows the 4-letter script codes.  "zh-Hans" versus "zh-Hant", for instance.  I find `script!("Hant")` *at least as* readable as `Script::TraditionalHan`.

> Actually, this might be a good question for the checklist: "Do some values have a different meaning / require special handling? If yes, lean towards enum." Country codes are a flat list, with no special cases; scripts have special cases, which makes it more likely that people will need to refer to those special cases directly, which makes the ergonomic issues more salient.

In my mental model, when identifiers are used, it is totally acceptable and even expected to have a certain subset of commonly used values available as constants.  For scripts, that could be perhaps the "special" script codes (Zyyy, Zzzz, ...) plus Latn (since Latn is often useful in computing applications).

> Also, the implicit values (Common, Inherited, Unknown) that developers are most likely to need to hardcode have particularly awful names (Zyyy, Zinh/Qaai, Zzzz). As a developer, I would much prefer to write (or read) `Script::Common` rather than `Script::from_code("Zyyy").unwrap()`.

I simply don't see the use case behind defining all 200+ script codes as constants *or* enums. If you are a developer who wants to work with a long-tail script code, then it's not difficult to ask that you write out the 4-letter script code.

> I'm not convinced that having to recompile for a new release is a sufficient problem. Given that ICU4C currently doesn't support swapping out data without recompiling, nobody's existing workflow will be blocked if ICU4X doesn't support that use case.

I have been trying to make the case since the inception of ICU4X that a core feature of ICU4X relative to ICU4C is the ability to decouple code version from data version.  We know of several Google clients where this is a major pain point in adopting ICU; they want to have a single data file shared across a whole app or platform, even if different pieces of the app or platform were compiled against different versions of the code.  It would be a *major* shift if ICU4X started imposing the code-data requirement, and I believe it would greatly diminish the value of ICU4X for a number of clients.

Comment (iainireland) [reply docs/design/enums_or_ids.md:10]: > Not sure if I agree with that. Anyone familiar with BCP-47 already knows the 4-letter script codes. "zh-Hans" versus "zh-Hant", for instance. I find script!("Hant") at least as readable as Script::TraditionalHan.
Neither "Hans" nor "Hant" is a valid script in this context. Han characters are assigned `Script::Han`, with four-letter code "Hani".

Comment (iainireland) [reply docs/design/enums_or_ids.md:10]: > I have been trying to make the case since the inception of ICU4X that a core feature of ICU4X relative to ICU4C is the ability to decouple code version from data version. We know of several Google clients where this is a major pain point in adopting ICU; they want to have a single data file shared across a whole app or platform, even if different pieces of the app or platform were compiled against different versions of the code. It would be a major shift if ICU4X started imposing the code-data requirement, and I believe it would greatly diminish the value of ICU4X for a number of clients.

Fair point. I retract my objection.

> In my mental model, when identifiers are used, it is totally acceptable and even expected to have a certain subset of commonly used values available as constants. For scripts, that could be perhaps the "special" script codes (Zyyy, Zzzz, ...) plus Latn (since Latn is often useful in computing applications).
I simply don't see the use case behind defining all 200+ script codes as constants or enums. If you are a developer who wants to work with a long-tail script code, then it's not difficult to ask that you write out the 4-letter script code.

I guess I see it the other way around: if we're going to make Latin available as a constant / enum value, then I don't see the harm in also making other scripts available. The 4-letter codes are obscure enough (and sufficiently inconsistent with uses elsewhere) that users are generally going to have to go look up the correct code. It doesn't cost us anything to make these values available.

I think it's useful to think about two cases: Script as input to ICU4X, and Script as output from ICU4X.

**Input**: If the user tries to create a Script to pass in to ICU4X, we have to validate it. 

In the identifier case, the user provides a 4-letter code. We look it up in a data-driven map and, if it's a valid script, create a 4-byte TinyStr wrapped in a newtype. With some proc macro trickery, we can do this at compile time for constants. One advantage here is that new scripts can be added to the data post-compilation, and we can still correctly parse the new script name and return the new data. 

In the enum case, constructing a known constant script is trivial. If the user needs to look up a dynamic value (most likely a string) then it needs to be converted to an enum first. We can provide a similar API that accepts 4-letter codes. Relative to identifiers, the disadvantage is that we can't accept codes that weren't known at compile time; a possible advantage is that we're matching against a bunch of known constants, so LLVM may be able to do a better job optimizing.

**Output**: If the user asks "what script does this character belong to", we have to be able to represent the answer.

In the identifier case, we can look up the integer value of the script in the codepoint trie, and then look up that integer value in a data-driven side table to determine the associated 4-letter code. The advantage here is that we can return a script value that was unknown at compile time.

In the enum case, we also look up the integer value of the script in the codepoint trie. If it's a value we recognize, we're done. If not, we have to decide what to do. We have several options here, but I think the best one is to just return `Script::Unknown`. Note that so long as scripts are allocated sequentially, I think this ends up being a single comparison. If the user wants the short name or long name, we can provide separate functions to look those up, equivalent to the side-table lookup that we always do in the identifier case.

So it seems to me that both approaches are capable of handling the code/data separation, and the advantages of the identifier version (better support for eg Klingon in codebases that were compiled before it was added to Unicode) are smaller than the ergonomic advantages of using enums, at least in the specific case of Script where new additions are expected to see incredibly marginal use.

(I'm starting to think there may also be a proc-macro based approach that gives us the data-driven forward-compatibility benefits of identifiers with the ergonomic advantages of enums, but I want to try writing some code before I make strong claims about that.)

Comment (sffc) [reply docs/design/enums_or_ids.md:10]: Fair points.

> I guess I see it the other way around: if we're going to make Latin available as a constant / enum value, then I don't see the harm in also making other scripts available. The 4-letter codes are obscure enough (and sufficiently inconsistent with uses elsewhere) that users are generally going to have to go look up the correct code. It doesn't cost us anything to make these values available.

I see the cost as being the ongoing human cost of having to update ICU4X APIs to conform to the latest Unicode standards; it puts our release process in the way of downstream consumers fetching the latest and greatest Unicode data, instead of allowing them to just re-generate data offline and use it with an older version of ICU4X.

> In the enum case, we also look up the integer value of the script in the codepoint trie. If it's a value we recognize, we're done. If not, we have to decide what to do. We have several options here, but I think the best one is to just return `Script::Unknown`.

This is a possibility.  It still makes the ICU4X dev team the bottleneck, because new APIs are required in order to interpret the new data.

*If* we were to use an enum, I think `Script::Unknown` should be a variant that wraps a `u16`, such that a data-driven map between script numbers and script codes or script names can still work correctly.

**Comment by sffc** [comment]
Okay, I "flattened" the decision tree to a list of five questions.  I think this should satisfy all of the feedback.  PTAL and feel free to merge once everyone has had a chance to review.

File: docs/design/enums_or_ids.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+```
+
+If an enum is used for interchange between programs, including over FFI, discriminants should be explicitly specified. However, if the enum is only used within the context of the current program (if it never leaves Rust), explicitly specifying the discriminants could limit the compiler's ability to optimize the enum.
+
+### Non-exhaustive enums
+
+It is possible in Rust to mark enums as `#[non_exhaustive]`, which allows additional entries to be added over time.
+
+However, if there is a need to do this, then it means that the set of entities could grow over time. If this is the case, you should probably be using identifiers instead.
```
Comment (zbraniecki) [inline docs/design/enums_or_ids.md:85]: this sentence seems to indicate that option bag values (which you wanted to have as non exhaustive) should be identifiers, which I disagree with.

**Review by zbraniecki — DISMISSED** [review]
lgtm except of the claim that using non-exhaustive on an enum is an indication that identifier should be used in its place. I disagree with that notion.

File: docs/design/enums_or_ids.md (hunk 3 above)
Comment (sffc) [reply docs/design/enums_or_ids.md:85]: Option bags are not sets of entities.  This doc is only about sets of entities.

Comment (sffc) [reply docs/design/enums_or_ids.md]: I weakened the statement.

Comment (zbraniecki) [reply docs/design/enums_or_ids.md]: Option bag options values may be:

```rust
#[non_exhaustive]
enum DateLength {
  Short,
  Medium,
  Long
}

struct Options {
  date: DateLength
}
```

Comment (zbraniecki) [reply docs/design/enums_or_ids.md]: > I weakened the statement.

The statement is still stronger than my opinion. It seems to nudge reader in the direction of not using enums for situations like in the example above where I claim enum to be a better choice than identifier.

I'm ok with merging this PR as is, I don't think we'll always agree on such topics so moving toward my position would likely start pushing this away from yours.

Comment (sffc) [reply docs/design/enums_or_ids.md]: Let's look at DateLength and put it on the 5 point decision tree.

1. Is the set **fixed**?
    - MAYBE; it grows fairly rarely
2. Is the set **comprehensive**? Is it inclusive of all peoples and cultures? Are there no additional entities that are not currently represented?
    - YES; it is defined by CLDR in order to cover all peoples and cultures
3. Are there **30 or fewer items** in the set?
    - YES
4. Are the entities used within the context of a single program, and **not normally interchanged** between different programs?
    - YES
5. Are entities in the set usually **compared** with one another, and rarely expressed in a standalone context?
    - YES

So you can see that DateLength clearly favors Enum, even though we applied `#[non_exhaustive]` out of an abundance of caution.

Do you have a concrete suggestion on how to improve this language further, or are you satisfied?

Comment (sffc) [reply docs/design/enums_or_ids.md]: okay I also weakened the paragraph again.  PTAL and approve if you like the doc.

**Review by zbraniecki — APPROVED** [review]
lgtm!

**Review by Manishearth — APPROVED** [review]

File: docs/design/enums_or_ids.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+This design doc helps explain how we think about entities in ICU4X and when to use enums versus identifiers to represent them.
+
+## Decision Tree
+
+If you need to decide whether to use an enum or an identifier for a certain set of entities in ICU4X, consider the following questions:
+
+1. Is the set **fixed**?
+2. Is the set **comprehensive**? Is it inclusive of all peoples and cultures? Are there no additional entities that are not currently represented?
+3. Are there **30 or fewer items** in the set?
```
Comment (iainireland) [inline docs/design/enums_or_ids.md:16]: I continue to think that the number of entries is just a proxy for the other questions, and not meaningful on its own, but this is not a hill I want to die on.

**Review by iainireland — APPROVED** [review]
I'm happy enough with where this ended up.

**Comment by Manishearth** [comment]
Shane's on PTO, going to go ahead and merge this
